server/src/test/java/io/crate/planner/UnionPlannerTest.java

```diff
@@ -194,4 +205,36 @@ public void test_union_with_different_types_in_queries_adds_casts() {
         ));
     }
 
+    @Test
+    public void test_union_estimated_rows_not_available() {
+        String stmt = "select id from users union all select id from locations";
+        TableStats tableStats = new TableStats();
+        Map<RelationName, Stats> rowCountByTable = new HashMap<>();
+        rowCountByTable.put(USER_TABLE_IDENT, new Stats(-1, 0, Map.of()));
+        rowCountByTable.put(TEST_DOC_LOCATIONS_TABLE_IDENT, new Stats(1, 0, Map.of()));
+        tableStats.updateTableStats(rowCountByTable);
+
+        var plan = createLogicalPlan(stmt, tableStats);
+        var union = (Union) plan.sources().get(0);
+        assertThat(union.numExpectedRows(), is(-1L));
+
+        rowCountByTable.put(USER_TABLE_IDENT, new Stats(1, 0, Map.of()));
+        rowCountByTable.put(TEST_DOC_LOCATIONS_TABLE_IDENT, new Stats(-1, 0, Map.of()));
+        tableStats.updateTableStats(rowCountByTable);
+
+        plan = createLogicalPlan(stmt, tableStats);
+        union = (Union) plan.sources().get(0);
+        assertThat(union.numExpectedRows(), is(-1L));
+    }
+
+    private LogicalPlan createLogicalPlan(String stmt, TableStats tableStats) {
+        PlannerContext context = e.getPlannerContext(clusterService.state());
+        LogicalPlanner logicalPlanner = new LogicalPlanner(
+            e.nodeCtx,
+            tableStats,
+            () -> clusterService.state().nodes().getMinNodeVersion()
+        );
+        return logicalPlanner.plan(e.analyze(stmt), context);
+    }
```

Comment:
minor suggestion, no strong opinion but I think we could inline the helper method while still keeping readability.
```suggestion
        var context = e.getPlannerContext(clusterService.state());
        var logicalPlanner = new LogicalPlanner(
            e.nodeCtx,
            tableStats,
            () -> clusterService.state().nodes().getMinNodeVersion()
        );
        var plan = logicalPlanner.plan(e.analyze(stmt), context);
        var union = (Union) plan.sources().get(0);
        assertThat(union.numExpectedRows(), is(-1L));

        rowCountByTable.put(USER_TABLE_IDENT, new Stats(1, 0, Map.of()));
        rowCountByTable.put(TEST_DOC_LOCATIONS_TABLE_IDENT, new Stats(-1, 0, Map.of()));
        tableStats.updateTableStats(rowCountByTable);

        plan = logicalPlanner.plan(e.analyze(stmt), context);
        union = (Union) plan.sources().get(0);
        assertThat(union.numExpectedRows(), is(-1L));
    }
```